### ISSUE TITLE
download file from google drive.

### ISSUE THREAD

**Issue by DexterFstone** [godotengine/godot#90160]
### Tested versions

reproducible in 4.2.1.stable, 3.5.3.stable

### System information

Windows 10 - Godot 4.2.1 Stable

### Issue description

I tried to download a file from Google Drive, the problem is if I paste my link to the browser it works but I don’t know why inside Godot is not working
The URL is valid and you can be tested, The download link is direct
```
extends Node2D

var url = "https://drive.google.com/uc?export=download&id=16oF7J-vGCGYTY6qIkyir4pro8qB-svLo"

@onready var http_request: HTTPRequest = $HTTPRequest

func _ready() -> void:
	if FileAccess.file_exists("user://game_config.ini"):
		var file = FileAccess.open("user://game_config.ini",FileAccess.READ)
	http_request.download_file = "user://game_config.ini"
	http_request.request(url)

func _on_http_request_request_completed(result: int, response_code: int, headers: PackedStringArray, body: PackedByteArray) -> void:
	print(result)
	print(response_code)
	print(headers)
	print(body)
```

### Steps to reproduce

Just copy the script and be ready to go

### Minimal reproduction project (MRP)

N/A

**Comment by AThousandShips** [comment]
Please check the error and status of the request, use:
```gdscript
var err = http_request.request(url)
if err != OK:
    print(error_string(err))
```

**Comment by timothyqiu** [comment]
The response is 303 See Other. `HTTPRequest` only handles 301 and 302.

Currently, you can get the actual URL from the Location header in a 303 response and request again.

**Comment by Calinou** [comment]
Reopening, as this is considered a bug that should be resolved.

**Comment by hbdsklf** [comment]
I have the same problem with 307 status codes. The problem appears to be here: https://github.com/godotengine/godot/blob/master/scene/main/http_request.cpp#L230C2-L230C53

I guess the following response codes should trigger a redirect:

- 301 Moved Permanently
- 302 Found (Previously "Moved temporarily")
- 303 See Other (This should perform a GET request even if the original request was POST)
- 305 Use Proxy
- 307 Temporary Redirect
- 308 Permanent Redirect

As a work-around you can do something like the following (I use this technique for now, but not with this exact code, so don't copy this code 1:1 and expect it to work out of the box. Feedback welcome.):

```
var status_code : int = 0
# This only fixes 307, more codes could be added
# Max number of redirects is not taken into account
while (status_code == 0 || status_code == 307):
	if (http_request.request(url, headers) != OK):
		break
	var result = await http_request.request_completed
	status_code = result[1]
```

**Comment by GrammAcc** [comment]
I'd like to take a shot at adding the missing redirects if no one is working on this yet.

I agree that 300 and 304 shouldn't need to be implemented. Also, 305 can be treated the same as 302 on the client-side, so I guess we can treat those as synonymous.

For the other redirects, I'll just implement the behavior according to [RFC9110](https://www.rfc-editor.org/rfc/rfc9110#name-status-codes)

I'm a Backend Developer by trade, but this is the first time I've looked at Godot's networking APIs, so if there is anything I should know about request handling in Godot that differs from the standards, please let me know. Thanks!

**Comment by hbdsklf** [comment]
Nice, thx! I would happily test your fix(es) against my use case. I've never built Godot from source so far, but the process seems to be quite easy and well documented, so I should manage. Tell me if you think this makes sense.

**Comment by GrammAcc** [comment]
> Nice, thx! I would happily test your fix(es) against my use case. I've never built Godot from source so far, but the process seems to be quite easy and well documented, so I should manage. Tell me if you think this makes sense.

That would be a big help, thanks! Not sure when I'll have something ready to test. I can usually only work on Godot on Saturdays, and I'm going to a local programming meetup today. But I'll ping you in this issue once I have something ready.

As for building Godot from source, it's one of the easiest build processes I've had to go through, but like a lot of large software, it can be slower than you might expect. On my machine (AMD Ryzen 5 2600, 6 cores, 12 threads), it takes about 30-45 minutes to build from scratch. That might vary based on the platform as well. I'm on `linuxbsd`. Just wanted to give you a heads up since it might seem like it's not working but it just takes a while. :)

**Comment by GrammAcc** [comment]
@hbdsklf I just opened #91261 for this one. :)

If you have the chance, can you test out https://github.com/GrammAcc/godot/commit/0811fffb23f03e69eb5bbc0c6fc324db69ecb171 against your use-case?

I added all the redirect codes except for 300 and 304.

307 and 308 should automatically redirect on any safe methods (GET, HEAD, OPTIONS, TRACE), but for unsafe methods such as POST or DELETE, they will not automatically redirect for security reasons.

301, 302, 303, and 305 will automatically redirect, but once #91199 is merged, unsafe methods will be changed to GET and the body will be discarded, so if you are relying on POST, DELETE, etc. being automatically redirected, that behavior will change soon. :)

CONNECT requests will never automatically redirect for any status code. :)

Let me know if you have any questions for me. Thanks!

**Comment by hbdsklf** [comment]
@GrammAcc I'll see what I can do in the upcoming days and get back to you here. Thanks for your work!

**Comment by webredraptors** [comment]
It looks like #91199 is set to be deployed in 4.3. But it seems #91261 has no milestone set. Will fixes for other 3XX HTTP responses be postponed to a later release?
